src/servers/ws-server.ts

```diff
@@ -45,26 +45,28 @@ const getContext = async (ctx: Context) => {
   const ipString = isProd
     ? connectionParams?.["x-real-ip"] || connectionParams?.["x-forwarded-for"]
     : connectionParams?.ip
+    ? connectionParams?.ip
+    : ctx.extra
+    ? (ctx.extra as Extra).request?.socket?.remoteAddress
```

Comment:
casting is generally a bad practice. why is that needed?

Reply:
yep, I was trying to get the e2e tests to pass and it kept failing with an IP address parsing error, so I committed this to see if it would pass. I cleaned up the code with a fully typed context and nullish coalescing operators.